pkg/k8s/client/cell.go

```diff
@@ -109,29 +112,69 @@ type compositeClientset struct {
 	log           logrus.FieldLogger
 	closeAllConns func()
 	restConfig    *rest.Config
+	rt            *rotatingHttpRoundTripper
+
+	lock.RWMutex
 }
 
 func newClientset(lc hive.Lifecycle, log logrus.FieldLogger, cfg Config) (Clientset, error) {
 	if !cfg.isEnabled() {
 		return &compositeClientset{disabled: true}, nil
 	}
 
-	if cfg.K8sAPIServer != "" &&
-		!strings.HasPrefix(cfg.K8sAPIServer, "http") {
-		cfg.K8sAPIServer = "http://" + cfg.K8sAPIServer // default to HTTP
+	apiServerURLs := []string{}
+
+	for _, apiServerURL := range cfg.K8sAPIServerURLs {
+		if apiServerURL == "" {
+			continue
+		}
+
+		if !strings.HasPrefix(apiServerURL, "http") {
+			apiServerURL = fmt.Sprintf("http://%s", apiServerURL)
+		}
+
+		_, err := url.Parse(apiServerURL)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to parse APIServerURL %s, skipping", apiServerURL)
+			continue
+		}
+
+		apiServerURLs = append(apiServerURLs, apiServerURL)
 	}
 
+	cfg.K8sAPIServerURLs = apiServerURLs
+
 	client := compositeClientset{
 		log:        log,
 		controller: controller.NewManager(),
 		config:     cfg,
+		rt:         &rotatingHttpRoundTripper{
+			log: log,
+		},
+
 	}
 
 	restConfig, err := createConfig(cfg.K8sAPIServer, cfg.K8sKubeConfigPath, cfg.K8sClientQPS, cfg.K8sClientBurst)
 	if err != nil {
 		return nil, fmt.Errorf("unable to create k8s client rest configuration: %w", err)
 	}
+
+	if client.canRotateAPIServerURL() {
+		if err := client.rotateAPIServerURL(); err != nil {
+			return nil, err
+		}
+	} else {
+		apiServerURL, err := url.Parse(restConfig.Host)
+		if err != nil {
+			return nil, err
+		}
+		client.rt.apiServerURL = apiServerURL
+		log.Infof("Setting APIServerURL to %s", apiServerURL)
+	}
+
+	restConfig.Wrap(client.WrapRoundTripper)
```

Comment:
This makes rotatingHttpRoundTripper.Roundtrip run before the pre-configured one. This enables us to change the receiving Host dynamically.